crates/ruma-events/src/direct.rs

```diff
@@ -7,23 +7,133 @@ use std::{
     ops::{Deref, DerefMut},
 };
 
-use ruma_common::{OwnedRoomId, OwnedUserId};
-use ruma_macros::EventContent;
+use ruma_common::{IdParseError, OwnedRoomId, OwnedUserId, UserId};
+use ruma_macros::{EventContent, IdZst};
 use serde::{Deserialize, Serialize};
 
+/// An user identifier, it can be a [`UserId`] or a third-party identifier
+/// like an email or a phone number.
+///
+/// There is no validation on this type, any string is allowed,
+/// but you can use `as_user_id` or `into_user_id` to try to get an [`UserId`].
+#[repr(transparent)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, IdZst)]
+pub struct DirectUserIdentifier(str);
+
+impl DirectUserIdentifier {
+    /// Get this `DirectUserIdentifier` as an &[`UserId`] if it is one.
+    pub fn as_user_id(&self) -> Option<&UserId> {
+        // TODO does it allocate ? Can we avoid it ?
```

Comment:
@zecakeh if you have an idea. If not, I will probably just remove the TODO.

Reply:
This does not allocate, you can remove the TODOs.